### ISSUE TITLE
Audio distortion in firefox

### ISSUE THREAD

**Issue by trichimtrich** [giongto35/cloud-game#179]
Ingame sound of `sushi the cat` is distortion in Firefox,
Checked with `retroarch` (same core+game), the stereo sound is quite different.
Output sample rate of GBA is 32768, and at browser is 48000/2 channels.
Can someone introduce how we resample and convert the sound stream.

**Comment by sergystepanov** [comment]
I have not touched that part yet, so better ask @giongto35. Sampling part should be alright, I guess.
What about [resampling](https://github.com/giongto35/cloud-game/blob/b7f47f5519661abd68e8353b496ae8a93c4589b2/pkg/worker/room/media.go#L15), since we are using Opus codec with max bandwidth, so it must have 48kHz frequency and thus audio resampling is a must. Actually, it's possible now to use raw PCM in WebRTC as far as I know.
I don't know about this distortion, it's Firefox specific, interesting why and how Chromium can handle it. SoX resampler is a standard, might try it out.

On the high level we take 16bit PCM encoded audio samples encoded as [L,R,L,R,L,R] from a core, then wait until a preconfigured frame duration buffer (of 10ms) will be full (an audio frame just contains a number of audio samples), then resample our samples with 48kHz (not sure about the algo, but it's just stretching of previous sample durations???) and that's it, I guess.

**Comment by sergystepanov** [comment]
Hm... with `const AUDIO_MS = 20` or 20ms frame duration it plays perfectly in FIrefox just as in Retroarch.
Not sure about [that statement](https://wiki.xiph.org/index.php?title=Opus_Recommended_Settings&mobileaction=toggle_view_desktop#Framesize_Tweaking) what does it mean.
In Chrome, actually, it is playing in mono for some reason...

**Comment by trichimtrich** [comment]
> Opus uses a 20 ms frame size by default, as it gives a decent mix of low latency and good quality.
> 
> Could you check if it's working for you? Then fix it like that and leave for better times. (:

Very nice finding 👍It sounds good now.

Here the opus decoder in firefox rtc, the config is constant (?)

https://github.com/mozilla/gecko-dev/blob/50b3fb522bdb080a7c9c00b1fdc758d171586cb6/media/webrtc/trunk/webrtc/modules/audio_coding/codecs/opus/opus_interface.c#L278-L285

```
    /* Create new memory, always at 48000 Hz. */
    state->decoder = opus_decoder_create(48000, (int)channels, &error);
    if (error == OPUS_OK && state->decoder != NULL) {
      /* Creation of memory all ok. */
      state->channels = channels;
      state->prev_decoded_samples = kWebRtcOpusDefaultFrameSize;
      state->in_dtx_mode = 0;
      *inst = state;
      return 0;
    }
```

```
/* Default frame size, 20 ms @ 48 kHz, in samples (for one channel). */
  kWebRtcOpusDefaultFrameSize = 960,
```

But the same decoder in chromium is slightly different, more dynamic way (?)

https://source.chromium.org/chromium/chromium/src/+/master:third_party/webrtc/modules/audio_coding/codecs/opus/opus_interface.cc;l=401-412?q=opus_decoder_create&ss=chromium&originalUrl=https:%2F%2Fcs.chromium.org%2F

The sterero -> mono you talked about make sense to me.

**Comment by sergystepanov** [comment]
Great, I'll change it then.
Speaking about frames, in any case it shouldn't matter for an opus decoder what frame size you are using, just input frequency. I can't tell why it's matter in our case, both decoders seem to me as identical, they should obey WebRTC Opus specification in any case. There is a lot is going under the hood there. IDK, this issue requires a ton of researching and learning, maybe I'll come back to it later if I got nothing better to do (:.

**Comment by sergystepanov** [comment]
Fixed for now in ae59586a3d9d2c4f4844c5f2ccc0a4105093ddc3.
